mypy/checkmember.py

```diff
@@ -228,17 +228,26 @@ def lookup_member_var_or_accessor(info: TypeInfo, name: str,
         return None
 
 
-def check_method_type(functype: FunctionLike, itype: Instance,
+def check_method_type(functype: FunctionLike, itype: Instance, is_classmethod: bool,
                       context: Context, msg: MessageBuilder) -> None:
     for item in functype.items():
         if not item.arg_types or item.arg_kinds[0] not in (ARG_POS, ARG_STAR):
             # No positional first (self) argument (*args is okay).
             msg.invalid_method_type(item, context)
-        else:
+        elif not is_classmethod:
             # Check that self argument has type 'Any' or valid instance type.
             selfarg = item.arg_types[0]
             if not subtypes.is_equivalent(selfarg, itype):
                 msg.invalid_method_type(item, context)
+        else:
+            # Check that cls argument has type 'Any' or valid class type.
+            clsarg = item.arg_types[0]
+            if isinstance(clsarg, CallableType) and clsarg.is_type_obj():
```

Comment:
Just checking the return type would not be sufficient in general, at least
if we are going to have `Type[...]` with explicit argument types that might
be incompatible with `__init__`. So if that's the case, we'd have to verify
that the argument types are compatible with the actual __init__ signature.
However, as that's not supported, it's not very useful to implement that
right now. So the current check seems reasonable, though it's probably
worth adding a comment about the potential future issue.

On Wed, Mar 23, 2016 at 2:41 PM, Guido van Rossum <notifications@github.com>
wrote:

> In mypy/checkmember.py
> <https://github.com/python/mypy/pull/1313#discussion_r57170202>:
>
> >              # Check that self argument has type 'Any' or valid instance type.
> >              selfarg = item.arg_types[0]
> >              if not subtypes.is_equivalent(selfarg, itype):
> >                  msg.invalid_method_type(item, context)
> > +        else:
> > +            # Check that cls argument has type 'Any' or valid class type.
> > +            clsarg = item.arg_types[0]
> > +            if isinstance(clsarg, CallableType) and clsarg.is_type_obj():
>
> I wasn't quite sure whether to require equivalence or a subclass
> relationship. Because the check for self requires equivalence I figured I'd
> do the same thing here. There's only one callsite (L190 above).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/python/mypy/pull/1313/files/f4cf5f4f33e22d9c626f5f04ea6194e71cb17077#r57170202>
>